### PR TITLE
[NTOS:MM] ARM3/sysldr.c: Fix string length handling

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

Be consistent: work right.

## Proposed changes

* Always set/use consistent Length and MaximumLength.
* Set intended UNICODE_NULL on a FullDllName.

### CODE REVIEW

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 1]
```diff
@@ -822,11 +822,9 @@ MiSnapThunk(IN PVOID DllBase,
             /* Now assume failure in case the forwarder doesn't exist */
             Status = STATUS_DRIVER_ENTRYPOINT_NOT_FOUND;
 
-            /* Build the forwarder name */
+            /* Build the forwarder name, including the dot */
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:825]: Why "including the dot"? The forwarder name is the name of the dll without the extension (i.e. without the first dot and what follows it) (Aside note: actually the extension should start at the latest dot, so I don't know whether using the strchr here is intended or not, i.e. whether windows does this or not).

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 2]
```diff
@@ -822,11 +822,9 @@ MiSnapThunk(IN PVOID DllBase,
...
             DllName.Buffer = (PCHAR)Address->u1.Function;
-            DllName.Length = (USHORT)(strchr(DllName.Buffer, '.') -
-                                      DllName.Buffer) +
-                                      sizeof(ANSI_NULL);
+            DllName.Length = (USHORT)(strchr(DllName.Buffer, '.') - DllName.Buffer + strlen("."));
```
Comment (HBelusca) [inline ntoskrnl/mm/ARM3/sysldr.c:827]: Wrong.

Comment (SergeGautherie) [reply ntoskrnl/mm/ARM3/sysldr.c:827]: What is wrong?

File: ntoskrnl/mm/ARM3/sysldr.c (hunk 1 above)
Comment (SergeGautherie) [reply ntoskrnl/mm/ARM3/sysldr.c:825]: I am not sure what your point is: I am simply documenting existing code (with regard to strings), which seems consistent as far as I looked.
If you think it does not behave as it should, file an issue...

File: ntoskrnl/mm/ARM3/sysldr.c (hunk 2 above)
Comment (tkreuzer) [reply ntoskrnl/mm/ARM3/sysldr.c:827]: Changing kernel code for the sake of changing it? using strlen here is retarded.

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 3]
```diff
@@ -2255,17 +2255,19 @@ MiInitializeLoadedModuleList(IN PLOADER_PARAMETER_BLOCK LoaderBlock)
         }
 
         /* Set the base name */
+        // NewEntry->BaseDllName.Length = LdrEntry->BaseDllName.Length; // Already.
+        NewEntry->BaseDllName.MaximumLength = NewEntry->BaseDllName.Length + sizeof(UNICODE_NULL);
         NewEntry->BaseDllName.Buffer = (PVOID)(NewEntry + 1);
 
-        /* Copy the full and base name */
+        /* Copy and null-terminate the full and base names */
```
Comment (tkreuzer) [inline ntoskrnl/mm/ARM3/sysldr.c:2262]: You seem to be determined to change as many lines in the kernel as possible. I really don't want git blame to one day tell me that you wrote the entire kernel.

File: ntoskrnl/mm/ARM3/sysldr.c [hunk 4]
```diff
@@ -3275,6 +3274,7 @@ MmLoadSystemImage(IN PUNICODE_STRING FileName,
             UnicodeTemp = PrefixName;
             UnicodeTemp.Buffer += 11;
             UnicodeTemp.Length -= (11 * sizeof(WCHAR));
+            UnicodeTemp.MaximumLength -= (11 * sizeof(WCHAR));
```
Comment (tkreuzer) [inline ntoskrnl/mm/ARM3/sysldr.c:3277]: Pointless.

**Review by tkreuzer — CHANGES_REQUESTED** [review]
I don't see anything in here that fixes anything, but plenty of random changes with the potential to cause subtle breakages.

**Comment by Extravert-ir** [comment]
The purpose of this PR is not clear: what does it fix?

**Pull request closed without merging**